sources/rest_api/client.py

```diff
@@ -0,0 +1,109 @@
+from typing import Optional, Dict, Any, Generator, Literal
+import copy
+
+from requests.auth import AuthBase
+from requests import Session as BaseSession
+
+
+from dlt.common import logger
+from dlt.sources.helpers.requests import client
+
+from .paginators import BasePaginator, UnspecifiedPaginator
+from .detector import create_paginator
+
+from .utils import join_url
+
+class RESTClient:
+    """A generic REST client for making requests to an API.
+
+    Attributes:
+        base_url (str): The base URL of the API.
+        headers (Optional[Dict[str, str]]): Headers to include in all requests.
+        auth (Optional[AuthBase]): An authentication object to use for all requests.
+        paginator (Optional[BasePaginator]): A paginator object for handling API pagination.
+            Note that this object will be deepcopied for each request to ensure that the
+            paginator state is not shared between requests.
+    """
+
+    def __init__(
+        self,
+        base_url: str,
+        headers: Optional[Dict[str, str]] = None,
+        auth: Optional[AuthBase] = None,
+        paginator: Optional[BasePaginator] = None,
+        session: BaseSession = None
+    ) -> None:
+        self.base_url = base_url
+        self.headers = headers
+        self.auth = auth
+        self.session = session or client.session
+        self.paginator = paginator if paginator else UnspecifiedPaginator()
+
+    def make_request(self, path="", method="get", params=None, json=None):
```

Comment:
we should IMO split this method into `_prepare_request` and `_make_request` and add a few "hook" methods for people to subclass easily.

in _prepare_request you create a new prepared request, pass it to authenticator and paginator for modification, IMO call an abstract method that people can implement for custom logic and return

in make request you do actual request